Support byte extract and update for new SMT backend

esteffin opened the pull request:
This PR adds support for byte_extract_exprt and byte_update_exprt for the new incremental SMT backend.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add regression tests for byte extraction/update

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -367,7 +367,7 @@ static smt_termt convert_expr_to_smt(
   const sub_expression_mapt &converted)
 {
-  if(operands_are_of_type<integer_bitvector_typet>(bitwise_and_expr))
+  if(operands_are_of_type<bitvector_typet>(bitwise_and_expr))
```
:pick: I know why this change is being made in `convert_expr_to_smt.cpp`. However the rationale is missing from the commit history and it isn't necessarily obvious why the change is needed. I suggest we split the change to this file into a separate commit. Then the change can be described in the commit title and the rationale explained in the commit body.
:hammer: The changes to `convert_expr_to_smt` are missing corresponding unit tests.

esteffin replied:
- Added unit tests.
- Moved to separate commit.

thomasspriggs commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -257,15 +258,17 @@ smt2_incremental_decision_proceduret::smt2_incremental_decision_proceduret(
     return;
   }
 
+  const exprt expr = lower_byte_operators(in_expr, ns);
```
:bulb: I suggest we rename `expr` to `lowered` or `lowered_expr`. `expr` is fine as a variable name for an `exprt` in a function using a single expression, but it becomes more confusing as soon as we have 2 of them.

esteffin replied:
Changed

thomasspriggs commented on src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -272,7 +272,7 @@ void smt2_incremental_decision_proceduret::ensure_handle_for_expr_defined(
   smt_define_function_commandt function{
     "B" + std::to_string(handle_sequence()), {}, convert_expr_to_smt(expr)};
-  expression_handle_identifiers.emplace(expr, function.identifier());
+  expression_handle_identifiers.emplace(in_expr, function.identifier());
```
:no_entry_sign: Please squash this commit into the previous commit with the other changes to `smt2_incremental_decision_procedure.cpp`. The problem this commit fixes is introduced in the previous commit and I dislike having bugs introduced and fixed in different commits within the same PR.

esteffin replied:
Squashed

thomasspriggs reviewed (commented):
This a partial review. I haven't reviewed the tests yet.

Commit:
Add lowering for byte extraction and update

When the input program includes non-aligned memory accesses the decision
procedure is passed expressions including byte_extract and byte_update
operators.
To support such programs in the decision procedure we need to lower such
operators to operators we support.

Commit:
Relax type of bitwise operations to any bitvector

Lowering of the byte extraction and byte update expressions can yield
non-integer bitvector typed operands.
Because we do not require a numeric interpretation of them we can relax
the requirement of integer-typed operands for bitwise operators.

thomasspriggs commented on unit/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -669,3 +669,230 @@ TEST_CASE(
+      make_test_symbol("extracted_byte", byte_type);
+    const auto original_int_symbol =
+      make_test_symbol("original_int", int64_type);
+    const auto byte_extract = make_byte_extract(
```
:no_entry_sign: This commit does not compile when checked out, due to a missing `#include <util/byte_operators.h>` needed for this line in the unit test. It is important that all commits at least compile and preferably that all enabled tests pass in order to make it possible to use `git bisect` for future problem finding.

thomasspriggs replied:
I think the problems with this particular test are just down to it being squashed into the wrong commit.

esteffin replied:
Fixed

thomasspriggs commented on unit/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -669,3 +669,230 @@ TEST_CASE(
+      smt_bit_vector_theoryt::make_or(
+        smt_bit_vector_theoryt::make_and(
+          original_int_term,
+          smt_bit_vector_constant_termt{18446744073709486335UL, 64}),
```
:pick: I think this mask would be easier to interpret as a hex literal.

esteffin replied:
Changed

thomasspriggs commented on unit/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -669,3 +669,230 @@ TEST_CASE(
+        smt_bit_vector_theoryt::make_and(
+          original_int_term,
+          smt_bit_vector_constant_termt{18446744073709486335UL, 64}),
+        smt_bit_vector_constant_termt{2816UL, 64}));
```
:pick: I think the literal `2816UL` may be easier to read as `11UL << 8`. This would make it clearer that it corresponds to the 11 in the input expression.

esteffin replied:
changed

thomasspriggs commented on unit/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp:
```diff
@@ -669,3 +669,230 @@ TEST_CASE(
   }
 }
+
+TEST_CASE(
```
:+1: Looks like a good test to me.

thomasspriggs reviewed (changes requested):
I think we should get the following sorted before merging -

1. Fixing the one unit test which is squashed into the wrong commit.
2. Adding commit message bodies. The commits now (mostly) say what they do, which is great. But some of the pre-requisite fixes are still missing the context as to why they are needed. I am happy to go through adding these with you.

Commit:
Create correct number of index id when with_exprt

As `with_exprt` can be used as a multy-ary operatior (with more than 3
arguments) we need to introduce an index smt_identifier_termt for each
index used by the expression.
